pkg/cloud/openstack/clients/secgroupservice.go

```diff
@@ -0,0 +1,358 @@
+			},
+			defaultRules...,
+		),
+	}
+}
+
+// matchGroups will check if security groups match.
+func (s *SecGroupService) matchGroups(desired, observed *openstackconfigv1.SecurityGroup) bool {
+	// If they have differing amount of rules they obviously don't match.
+	if len(desired.Rules) != len(observed.Rules) {
+		return false
+	}
+
+	// Rules aren't in any order, so we're doing this the hard way.
+	for _, desiredRule := range desired.Rules {
+		r := desiredRule
+		if r.RemoteGroupID == "self" {
+			r.RemoteGroupID = observed.ID
+		}
+		ruleMatched := false
+		for _, observedRule := range observed.Rules {
+			if s.matchRule(r, observedRule) {
+				ruleMatched = true
+				break
+			}
+		}
+
+		if !ruleMatched {
+			return false
+		}
+	}
+	return true
+}
+
+// reconcileGroup reconciles an already existing observed group by essentially emptying out all the rules and
+// recreating them.
+func (s *SecGroupService) reconcileGroup(desired, observed *openstackconfigv1.SecurityGroup) (*openstackconfigv1.SecurityGroup, error) {
+	klog.V(6).Infof("Deleting all rules for group %s", observed.Name)
+	for _, rule := range observed.Rules {
+		klog.V(6).Infof("Deleting rule %s from group %s", rule.ID, observed.Name)
+		err := rules.Delete(s.client, rule.ID).ExtractErr()
+		if err != nil {
+			return &openstackconfigv1.SecurityGroup{}, err
+		}
+	}
+	recreatedRules := make([]openstackconfigv1.SecurityGroupRule, len(desired.Rules), len(desired.Rules))
+	klog.V(6).Infof("Recreating all rules for group %s", observed.Name)
+	for _, rule := range desired.Rules {
+		r := rule
+		r.SecurityGroupID = observed.ID
+		if r.RemoteGroupID == "self" {
+			r.RemoteGroupID = observed.ID
+		}
+		newRule, err := s.createRule(r)
+		if err != nil {
+			return &openstackconfigv1.SecurityGroup{}, err
+		}
+		recreatedRules = append(recreatedRules, newRule)
+	}
+	observed.Rules = recreatedRules
+	return observed, nil
+}
+
+func (s *SecGroupService) createSecGroup(group openstackconfigv1.SecurityGroup) (*openstackconfigv1.SecurityGroup, error) {
+	createOpts := groups.CreateOpts{
+		Name:        group.Name,
+		Description: "Cluster API managed group",
+	}
+	klog.V(6).Infof("Creating group %+v", createOpts)
+	g, err := groups.Create(s.client, createOpts).Extract()
+	if err != nil {
+		return &openstackconfigv1.SecurityGroup{}, err
+	}
+
+	newGroup := s.convertOSSecGroupToConfigSecGroup(*g)
+	rules := make([]openstackconfigv1.SecurityGroupRule, len(group.Rules), len(group.Rules))
+	klog.V(6).Infof("Creating rules for group %s", group.Name)
+	for _, rule := range group.Rules {
+		r := rule
+		r.SecurityGroupID = newGroup.ID
+		if r.RemoteGroupID == "self" {
+			r.RemoteGroupID = newGroup.ID
+		}
+		newRule, err := s.createRule(r)
+		if err != nil {
+			return &openstackconfigv1.SecurityGroup{}, err
+		}
+		rules = append(rules, newRule)
+	}
+	newGroup.Rules = rules
+
+	return newGroup, nil
+}
+
+func (s *SecGroupService) matchRule(r1, r2 openstackconfigv1.SecurityGroupRule) bool {
```

Comment:
Would it make sense to implement this as `Equal` method on `SecurityGroupRule`? The standard library uses this pattern.
* https://golang.org/pkg/time/#Time.Equal
* https://golang.org/pkg/net/#IP.Equal